### PR TITLE
Feature/authenticate user

### PR DESCRIPTION
**Pull request by daudi13**
Hello reviewers,

In this PR I:

- Added authentication for user
- Activated cors
- Wrote test for authentication
- Added API routes for the login page 

---
thank you for taking the time to review my code.

### CODE REVIEW

File: app/services/authentication_token_service.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  def self.call(user_id)
+    payload = { user_id: }
+
+    JWT.encode payload, HMAC_SECRET, ALGORITHM_TYPE
+  end
+
+  def self.decode(token)
+    JWT.decode token, HMAC_SECRET, true, { algorithm: ALGORITHM_TYPE }[0]
+  end
+end
```
Comment (Beardless-sheik) [inline app/services/authentication_token_service.rb]: When I try to validate the token using the 'GET' /authenticate, I get this error

image

I passed the token as a header 'Authorization'. I checked the controller, and it does register and pick up the token but it seems the decode function in the AuthenticationServiceGives this error. Are you able to look into it? 

image

**Review by Beardless-sheik — COMMENTED** [review]
Hi @daudi13 Nice Job on the JWT auth!!!

Highlights:
- The 'POST' /api/v1/authenticate works great 
- Nice work using the bycrypt and JWT gem to authorise tokens and create password security on the user
- Nice work testing requests and your created AuthService
- No code style linter errors!

I did have one problem with the validateToken end point. Please check the inline comment

File: app/services/authentication_token_service.rb (hunk 1 above)
Comment (daudi13) [reply app/services/authentication_token_service.rb]: thank you @Beardless-sheik for pointing that out, I'll work on it in due time

File: app/controllers/api/v1/authentication_controller.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    raise AuthenticationError unless user.authenticate(params.require(:password))
+
+    token = AuthenticationTokenService.call(user.id)
+
+    render json: { token: }, status: :created
+  end
+
+  def token_validation
+    # Authorization: Bearer <token>
+    token, _options = token_and_options(request)
```
Comment (Beardless-sheik) [inline app/controllers/api/v1/authentication_controller.rb]: I get an error from the authentication service which seems to be coming from this line: 
- "#<JWT::DecodeError: Nil JSON web token>"

image


The token in line 19 gives nill. I think the previous way you used to grab the token: 

` token = request.headers['Authorization'] `
was a better way of picking the token as I see you solved the previous error in the Auth service

**Review by Beardless-sheik — COMMENTED** [review]
Hi, @daudi13 Great work fixing the Auth service but we still have one more pending error. 

Please see the inline comment. It should be an easy fix.

File: app/controllers/api/v1/authentication_controller.rb (hunk 2 above)
Comment (daudi13) [reply app/controllers/api/v1/authentication_controller.rb]: fixed... :+1:

**Review by Beardless-sheik — APPROVED** [review]
@daudi13 Great work. 

The GET /api/v1/authenticate returns authorized now with the correct header!

I believe we can merge it now :100:

**Comment by daudi13** [comment]
@Beardless-sheik  thanks man

**Pull request merged**